cmd/promtool/analyze.go

```diff
@@ -0,0 +1,488 @@
+
+// Query the related count_over_time(*_sum[lookback]) series to double check that metricName is a
+// histogram. This keeps the result small (avoids buckets) and the count gives scrape interval
+// when dividing lookback with it.
+func queryLabelSets(ctx context.Context, api v1.API, metricName string, end time.Time, lookback time.Duration) (model.Vector, error) {
+	query := fmt.Sprintf("count_over_time(%s{}[%s])", metricName, lookback.String())
+
+	values, _, err := api.Query(ctx, query, end)
+	if err != nil {
+		return nil, err
+	}
+
+	vector, ok := values.(model.Vector)
+	if !ok {
+		return nil, fmt.Errorf("query for metrics resulted in non-Vector")
+	}
+	return vector, nil
+}
+
+func seriesSelector(bucketMetricName string, lbs model.LabelSet) string {
+	builder := strings.Builder{}
+	builder.WriteString(bucketMetricName)
+	builder.WriteRune('{')
+	first := true
+	for l, v := range lbs {
+		if first {
+			first = false
+		} else {
+			builder.WriteRune(',')
+		}
+		builder.WriteString(string(l))
+		builder.WriteString("=\"")
+		builder.WriteString(string(v))
+		builder.WriteRune('"')
+	}
+	builder.WriteRune('}')
+
+	return builder.String()
+}
+
+func querySamples(ctx context.Context, api v1.API, query string, start, end time.Time, step time.Duration) (model.Matrix, error) {
+	values, _, err := api.QueryRange(ctx, query, v1.Range{Start: start, End: end, Step: step})
+	if err != nil {
+		return nil, err
+	}
+
+	matrix, ok := values.(model.Matrix)
+	if !ok {
+		return nil, fmt.Errorf("query of buckets resulted in non-Matrix")
+	}
+
+	return matrix, nil
+}
+
+type statistics struct {
+	min, max, available int
+	avg                 float64
+	scrapeInterval      time.Duration
+}
+
+func (s statistics) String() string {
+	return fmt.Sprintf("Bucket min/avg/max/avail@scrape: %d/%.3f/%d/%d@%v", s.min, s.avg, s.max, s.available, s.scrapeInterval)
+}
+
+type calcBucketStatisticsFunc func(matrix model.Matrix, step time.Duration, histo *statshistogram) (*statistics, error)
+
+func calcClassicBucketStatistics(matrix model.Matrix, step time.Duration, histo *statshistogram) (*statistics, error) {
+	numBuckets := len(matrix)
+
+	stats := &statistics{
+		min:            math.MaxInt,
+		available:      numBuckets,
+		scrapeInterval: step,
+	}
+
+	if numBuckets == 0 || len(matrix[0].Values) < 2 {
+		// Not enough data.
+		return stats, nil
+	}
+
+	numSamples := len(matrix[0].Values)
+
+	sortMatrix(matrix)
+
+	prev, err := getBucketCountsAtTime(matrix, numBuckets, 0)
+	if err != nil {
+		return stats, err
+	}
+
+	sumBucketsChanged := 0
+	for timeIdx := 1; timeIdx < numSamples; timeIdx++ {
+		curr, err := getBucketCountsAtTime(matrix, numBuckets, timeIdx)
+		if err != nil {
+			return stats, err
+		}
+
+		countBucketsChanged := 0
+		for bIdx := range matrix {
+			if curr[bIdx] > prev[bIdx] {
+				countBucketsChanged++
+			} else if curr[bIdx] < prev[bIdx] {
+				countBucketsChanged++
+				fmt.Printf("unexpected decrease in bucket %d at time %d\n", bIdx, timeIdx)
```

Comment:
This could legitimately happen for a gauge histogram (which also exists with classic histograms) or whenever a counter reset happens.